contracts/lib/LibPrices.sol

```diff
@@ -94,45 +111,41 @@ library Prices {
         PriceInstant[24] memory tracerPrices,
         PriceInstant[24] memory oraclePrices
     ) public pure returns (TWAP memory) {
-        uint256 instantDerivative = 0;
+        uint256 totalDerivativeTimeWeight = 0;
+        uint256 totalUnderlyingTimeWeight = 0;
         uint256 cumulativeDerivative = 0;
-        uint256 instantUnderlying = 0;
         uint256 cumulativeUnderlying = 0;
 
         for (uint256 i = 0; i < 8; i++) {
             uint256 currTimeWeight = 8 - i;
             // if hour < i loop back towards 0 from 23.
             // otherwise move from hour towards 0
-            uint256 j = hour < i ? 23 - i + hour : hour - i;
+            uint256 j = hour < i ? 24 - i + hour : hour - i;
 
             uint256 currDerivativePrice = averagePrice(tracerPrices[j]);
             uint256 currUnderlyingPrice = averagePrice(oraclePrices[j]);
 
-            if (currDerivativePrice > 0) {
-                instantDerivative += currTimeWeight;
+            // dont inclue periods that have no trades
+            if (currDerivativePrice == 0 && tracerPrices[j].trades == 0) {
+                continue;
+            } else {
+                totalDerivativeTimeWeight += currTimeWeight;
                 cumulativeDerivative += currTimeWeight * currDerivativePrice;
             }
 
-            if (currUnderlyingPrice > 0) {
-                instantUnderlying += currTimeWeight;
-                cumulativeUnderlying += currTimeWeight * currUnderlyingPrice;
-            }
-
-            if (instantDerivative == 0) {
-                return TWAP(0, 0);
+            // dont inclue periods that have no trades
+            if (currUnderlyingPrice == 0 && oraclePrices[j].trades == 0) {
```

Comment:
`oraclePrices[j].trades` is always going to be the same as `tracerPrices[j].trades` so the latter check is redundant

Reply:
If we consider each of these functions pure (eg can take any input) then this isn't strictly true. I'd rather be cautious than make use case specific assumptions.